langserver/workspace_refs.go

```diff
@@ -23,7 +24,18 @@ import (
 	"github.com/sourcegraph/jsonrpc2"
 )
 
+// WorkspaceReferencesTimeout is the timeout used for workspace/xreferences
+// calls. It is exposed so that external users of the language server may use
+// the same identical timeout in client requests. It cannot be entirely
+// enforced by the go-langserver, because go/loader doesn't support context
+// timeouts during lengthy type-checking operations.
+const WorkspaceReferencesTimeout = 15 * time.Second
+
 func (h *LangHandler) handleWorkspaceReferences(ctx context.Context, conn JSONRPC2Conn, req *jsonrpc2.Request, params lspext.WorkspaceReferencesParams) ([]referenceInformation, error) {
+	// TODO: Find a better way to pass this timeout down as part of the request
+	// via LSP itself.
+	ctx, cancel := context.WithTimeout(ctx, WorkspaceReferencesTimeout)
+	defer cancel()
```

Comment:
We should add support for https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#cancelRequest

I've already seen vscode send it to cancel symbol requests. This seems like the sort of thing that should be controlled by the caller. In the meantime though, this is sufficient.